core/variant/variant.cpp

```diff
@@ -3492,6 +3493,72 @@ bool Variant::hash_compare(const Variant &p_variant, int recursion_count) const
 	}
 }
 
+bool Variant::reference_compare(const Variant &p_variant) const {
+	if (type != p_variant.type) {
+		return false;
+	}
+
+	switch (type) {
+		case ARRAY: {
+			const Array &l = *(reinterpret_cast<const Array *>(_data._mem));
+			const Array &r = *(reinterpret_cast<const Array *>(p_variant._data._mem));
+			return l.id() == r.id();
+		} break;
+
+		case DICTIONARY: {
+			const Dictionary &l = *(reinterpret_cast<const Dictionary *>(_data._mem));
+			const Dictionary &r = *(reinterpret_cast<const Dictionary *>(p_variant._data._mem));
+			return l.id() == r.id();
+		} break;
+
+		case PACKED_INT32_ARRAY: {
+			const Vector<int32_t> &l = PackedArrayRef<int32_t>::get_array(_data.packed_array);
+			const Vector<int32_t> &r = PackedArrayRef<int32_t>::get_array(p_variant._data.packed_array);
+			return l.ptr() == r.ptr();
+		} break;
+
+		case PACKED_INT64_ARRAY: {
+			const Vector<int64_t> &l = PackedArrayRef<int64_t>::get_array(_data.packed_array);
+			const Vector<int64_t> &r = PackedArrayRef<int64_t>::get_array(p_variant._data.packed_array);
+			return l.ptr() == r.ptr();
+		} break;
+
+		case PACKED_FLOAT32_ARRAY: {
+			const Vector<float_t> &l = PackedArrayRef<float_t>::get_array(_data.packed_array);
+			const Vector<float_t> &r = PackedArrayRef<float_t>::get_array(p_variant._data.packed_array);
+			return l.ptr() == r.ptr();
+		} break;
+
+		case PACKED_FLOAT64_ARRAY: {
+			const Vector<double_t> &l = PackedArrayRef<double_t>::get_array(_data.packed_array);
+			const Vector<double_t> &r = PackedArrayRef<double_t>::get_array(p_variant._data.packed_array);
+			return l.ptr() == r.ptr();
+		} break;
+
+		case PACKED_VECTOR2_ARRAY: {
+			const Vector<Vector2> &l = PackedArrayRef<Vector2>::get_array(_data.packed_array);
+			const Vector<Vector2> &r = PackedArrayRef<Vector2>::get_array(p_variant._data.packed_array);
+			return l.ptr() == r.ptr();
+		} break;
+
+		case PACKED_VECTOR3_ARRAY: {
+			const Vector<Vector3> &l = PackedArrayRef<Vector3>::get_array(_data.packed_array);
+			const Vector<Vector3> &r = PackedArrayRef<Vector3>::get_array(p_variant._data.packed_array);
+			return l.ptr() == r.ptr();
+		} break;
+
+		case PACKED_COLOR_ARRAY: {
+			const Vector<Color> &l = PackedArrayRef<Color>::get_array(_data.packed_array);
+			const Vector<Color> &r = PackedArrayRef<Color>::get_array(p_variant._data.packed_array);
+			return l.ptr() == r.ptr();
+		} break;
+
+		default: {
+			return hash_compare(p_variant);
```

Comment:
Wanted to point out that equality shouldn't be determined based on the hashes but... it seems `hash_compare` is not comparing by hashes in the first place. :open_mouth: It's confusing, this method should be renamed (likely out of scope of this PR though).

```gdscript
var v1 := Vector2i(0, -117)
var v2 := Vector2i(884, -79)
prints(hash(v1), hash(v2), is_same(v1, v2)) # 1413829362 1413829362 false
```

Reply:
Yep, encountered multiple people that had expectation that value comparison happens by hash (and not by recursively going into each element of each type) so it is fast.

Reply:
Maybe `value_compare`, as opposed to the newly added `identity_compare`?